lib/rust/ensogl/core/src/display/object/class.rs

```diff
@@ -619,32 +694,168 @@ pub struct Id(usize);
 /// to be assigned to a particular [`scene::Layer`], and thus allows for easy to use depth
 /// management.
 #[derive(Derivative)]
-#[derive(CloneRef)]
-#[derivative(Clone(bound = ""), Default(bound = ""))]
-pub struct Instance<Host = Scene> {
+#[derive(CloneRef, Deref)]
+#[derivative(Clone(bound = ""))]
+pub struct Instance<Host: 'static = Scene> {
     rc: Rc<Model<Host>>,
 }
 
-impl<Host> Deref for Instance<Host> {
-    type Target = Rc<Model<Host>>;
-    fn deref(&self) -> &Self::Target {
-        &self.rc
-    }
-}
 
 impl<Host> Instance<Host> {
-    /// Constructor.
-    pub fn new() -> Self {
-        default()
-    }
-
     /// Create a new weak pointer to this display object instance.
     pub fn downgrade(&self) -> WeakInstance<Host> {
         let weak = Rc::downgrade(&self.rc);
         WeakInstance { weak }
     }
 }
 
+impl<Host: 'static> Instance<Host> {
+    /// Constructor.
+    pub fn new() -> Self
+    where Host: 'static {
+        Self { rc: Rc::new(Model::new()) }.init()
+    }
+
+    fn init(self) -> Self
+    where Host: 'static {
+        let network = &self.network;
+        let this = &self;
+        frp::extend! { network
+            eval self.event_source ((event) this.emit_event_impl(event));
+        }
+        self
+    }
+
+    fn new_event<T>(&self, payload: T) -> event::SomeEvent
+    where
+        Host: 'static,
+        T: 'static, {
+        event::SomeEvent::new(Some(self), payload)
+    }
+
+    fn _emit_event<T>(&self, payload: T)
+    where
+        Host: 'static,
+        T: 'static, {
+        self.event_source.emit(event::SomeEvent::new(Some(self), payload));
+    }
+
+    fn emit_event_impl(&self, event: &event::SomeEvent) {
+        let rev_parent_chain = self.rev_parent_chain();
+        if event.captures.get() {
+            for object in &rev_parent_chain {
+                object.capturing_event_fan.emit(&event.data);
+                if event.state() == event::State::Cancelled {
+                    break;
```

Comment:
I changed the implementation, now we can't return there. Also, even if we could, I really dislike early returns from functions. This is probably because of my Haskell background though ;)